From abc8837061c386997f485f4a360efaed73882570 Mon Sep 17 00:00:00 2001
From: Michel JAOUEN <michel.jaouen@st.com>
Date: Tue, 4 Feb 2025 19:44:43 +0100
Subject: [PATCH 1562/1695] plat: stm32mp2: add psa support on rse_comms

Use lib rse relying on ipcc mailbox and share memory,
Implement psa crypto api relying on mbedtls 3.6.2 include file for
reaching an rse comms host with the same mbedtls version

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git 2b1bf7088a37d46d430ec80530afa32e19c387f5 plat: stm32mp2: add psa support on rse_comms]
Change-Id: Ic910015336dae49daa3000d8faba54248b2f3618
Signed-off-by: Michel JAOUEN <michel.jaouen@st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/optee/optee_os/+/434431
Domain-Review: Thomas BOURGOIN <thomas.bourgoin@foss.st.com>
Reviewed-by: Thomas BOURGOIN <thomas.bourgoin@foss.st.com>
---
 core/arch/arm/plat-stm32mp2/conf.mk           |    5 +
 core/arch/arm/plat-stm32mp2/platform_config.h |    2 +
 .../arch/arm/plat-stm32mp2/psa_manifest/sid.h |   55 +
 .../plat-stm32mp2/psa_service/psa_client.c    |  121 ++
 .../arch/arm/plat-stm32mp2/psa_service/sub.mk |    3 +
 .../plat-stm32mp2/psa_service/tfm_crypto.c    | 1641 +++++++++++++++++
 .../psa_service/tfm_crypto_defs.h             |  187 ++
 core/arch/arm/plat-stm32mp2/sub.mk            |    1 +
 .../arm/plat-stm32mp2/tfm_builtin_key_ids.h   |   29 +
 9 files changed, 2044 insertions(+)
 create mode 100644 core/arch/arm/plat-stm32mp2/psa_manifest/sid.h
 create mode 100644 core/arch/arm/plat-stm32mp2/psa_service/psa_client.c
 create mode 100644 core/arch/arm/plat-stm32mp2/psa_service/sub.mk
 create mode 100644 core/arch/arm/plat-stm32mp2/psa_service/tfm_crypto.c
 create mode 100644 core/arch/arm/plat-stm32mp2/psa_service/tfm_crypto_defs.h
 create mode 100644 core/arch/arm/plat-stm32mp2/tfm_builtin_key_ids.h

diff --git a/core/arch/arm/plat-stm32mp2/conf.mk b/core/arch/arm/plat-stm32mp2/conf.mk
index a146d6a94..626213ec9 100644
--- a/core/arch/arm/plat-stm32mp2/conf.mk
+++ b/core/arch/arm/plat-stm32mp2/conf.mk
@@ -102,6 +102,7 @@ $(call force,CFG_STM32_TAMP,n)
 $(call force,CFG_STM32MP_PROVISIONING,n)
 $(call force,CFG_TA_STM32MP_NVMEM,n)
 $(call force,CFG_STM32MP2_CLK_CAL,n)
+$(call force,CFG_STM32_PSA_SERVICE,y)
 endif
 
 $(call force,CFG_ARM_GIC_PM,y)
@@ -361,6 +362,10 @@ ifeq ($(call cfg-one-enabled, CFG_STM32_RISAF CFG_STM32_RIFSC),y)
 $(call force,CFG_DRIVERS_FIREWALL,y)
 endif
 
+ifeq ($(CFG_STM32_PSA_SERVICE),y)
+$(call force,CFG_RSE_COMMS,y)
+endif
+
 # Trusted User Interface
 ifeq ($(CFG_WITH_TUI),y)
 $(call force,CFG_DISPLAY,y)
diff --git a/core/arch/arm/plat-stm32mp2/platform_config.h b/core/arch/arm/plat-stm32mp2/platform_config.h
index 54263db6e..2b5770a26 100644
--- a/core/arch/arm/plat-stm32mp2/platform_config.h
+++ b/core/arch/arm/plat-stm32mp2/platform_config.h
@@ -130,6 +130,8 @@
 
 #define OTP_MAX_SIZE			(STM32MP2_OTP_MAX_ID + 1U)
 
+#define PLAT_RSE_COMMS_PAYLOAD_MAX_SIZE (0x40 + 0x800)
+
 #define TARGET_CPU0_GIC_MASK		BIT(0)
 #define TARGET_CPU1_GIC_MASK		BIT(1)
 #define TARGET_CPUS_GIC_MASK		GENMASK_32(CFG_TEE_CORE_NB_CORE - 1, 0)
diff --git a/core/arch/arm/plat-stm32mp2/psa_manifest/sid.h b/core/arch/arm/plat-stm32mp2/psa_manifest/sid.h
new file mode 100644
index 000000000..5e1522924
--- /dev/null
+++ b/core/arch/arm/plat-stm32mp2/psa_manifest/sid.h
@@ -0,0 +1,55 @@
+/* SPDX-License-Identifier: BSD-3-Clause */
+/*
+ * Copyright (c) 2019-2022, Arm Limited. All rights reserved.
+ *
+ */
+
+/***********  WARNING: This is an auto-generated file. Do not edit!  **********/
+
+#ifndef __PSA_MANIFEST_SID_H__
+#define __PSA_MANIFEST_SID_H__
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+	/******** RSE_NS_MAILBOX_AGENT ********/
+#define RSE_MBOX_SERVICE_SID                                       (0x00000F00U)
+#define RSE_MBOX_SERVICE_VERSION                                   (1U)
+#define RSE_MBOX_SERVICE_HANDLE                                    (0x40000104U)
+
+	/******** RSE_SP_SCP ********/
+#define RSE_SCP_SERVICE_SID                                        (0x000000F0U)
+#define RSE_SCP_SERVICE_VERSION                                    (1U)
+#define RSE_SCP_SERVICE_HANDLE                                     (0x40000106U)
+
+	/******** RSE_SP_PS ********/
+#define RSE_PROTECTED_STORAGE_SERVICE_SID                          (0x00000060U)
+#define RSE_PROTECTED_STORAGE_SERVICE_VERSION                      (1U)
+#define RSE_PROTECTED_STORAGE_SERVICE_HANDLE                       (0x40000101U)
+
+	/******** RSE_SP_ITS ********/
+#define RSE_INTERNAL_TRUSTED_STORAGE_SERVICE_SID                   (0x00000070U)
+#define RSE_INTERNAL_TRUSTED_STORAGE_SERVICE_VERSION               (1U)
+#define RSE_INTERNAL_TRUSTED_STORAGE_SERVICE_HANDLE                (0x40000102U)
+
+	/******** RSE_SP_CRYPTO ********/
+#define RSE_CRYPTO_SID                                             (0x00000080U)
+#define RSE_CRYPTO_VERSION                                         (1U)
+#define RSE_CRYPTO_HANDLE                                          (0x40000100U)
+
+	/******** RSE_SP_PLATFORM ********/
+#define RSE_PLATFORM_SERVICE_SID                                   (0x00000040U)
+#define RSE_PLATFORM_SERVICE_VERSION                               (1U)
+#define RSE_PLATFORM_SERVICE_HANDLE                                (0x40000105U)
+
+	/******** RSE_SP_INITIAL_ATTESTATION ********/
+#define RSE_ATTESTATION_SERVICE_SID                                (0x00000020U)
+#define RSE_ATTESTATION_SERVICE_VERSION                            (1U)
+#define RSE_ATTESTATION_SERVICE_HANDLE                             (0x40000103U)
+
+#ifdef __cplusplus
+}
+#endif
+
+#endif /* __PSA_MANIFEST_SID_H__ */
diff --git a/core/arch/arm/plat-stm32mp2/psa_service/psa_client.c b/core/arch/arm/plat-stm32mp2/psa_service/psa_client.c
new file mode 100644
index 000000000..94234213f
--- /dev/null
+++ b/core/arch/arm/plat-stm32mp2/psa_service/psa_client.c
@@ -0,0 +1,121 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright (c) 2025, STMicroelectronics
+ */
+
+#include <drivers/mailbox.h>
+#include <drivers/stm32mp_dt_bindings.h>
+#include <initcall.h>
+#include <kernel/dt.h>
+#include <kernel/misc.h>
+#include <libfdt.h>
+#include <mm/core_memprot.h>
+#include <rse_comms.h>
+#include <trace.h>
+#include <util.h>
+
+#define ST_PSA_CLIENT_COMPAT	"st,psa-client"
+
+/* Arbitrary default Client */
+#define ST_PSA_CLIENT_ID (-19)
+
+static struct mbox_chan *handle_send_receive;
+
+struct stm32_rse_shmem {
+	uint32_t size;
+	uint8_t payload[];
+};
+
+static struct stm32_rse_shmem *shmem;
+static size_t shmem_size;
+
+static int rse_mbox_send_data(const uint8_t *send_buffer, size_t size)
+{
+	TEE_Result res = TEE_ERROR_NO_DATA;
+
+	memcpy(shmem->payload, send_buffer, size);
+	shmem->size = size;
+	res = mbox_send(handle_send_receive, false, NULL, 0);
+	if (res != TEE_SUCCESS) {
+		EMSG("Send/receive IPCC : send failed");
+		return res;
+	}
+	return 0;
+}
+
+static int rse_mbox_rcv_data(const uint8_t *rcv_buffer, size_t *size)
+{
+	TEE_Result res = TEE_ERROR_NO_DATA;
+
+	do {
+		res = mbox_recv(handle_send_receive, false, NULL, 0);
+	} while (res == TEE_ERROR_NO_DATA);
+	if (res != TEE_SUCCESS) {
+		EMSG("Send/receive : failed to receive message");
+		return res;
+	}
+	*size = shmem->size;
+	memcpy((void *)rcv_buffer, shmem->payload, *size);
+	return 0;
+}
+
+static int rse_mbox_size_data(size_t  *size)
+{
+	*size = shmem_size;
+	if (!shmem_size)
+		return -1;
+	else
+		return 0;
+}
+
+static TEE_Result stm32mp_start_psa_service(void)
+{
+	TEE_Result res = TEE_SUCCESS;
+	void *fdt = NULL;
+	const fdt32_t *prop = NULL;
+	int node = 0;
+	int len = 0;
+	int pnode = 0;
+	paddr_t pa = 0;
+
+	fdt = get_embedded_dt();
+	node = fdt_node_offset_by_compatible(fdt, 0, ST_PSA_CLIENT_COMPAT);
+	if (node < 0)
+		/*  Test not activated in this dt configuration */
+		return TEE_SUCCESS;
+
+	if (fdt_get_status(fdt, node) == DT_STATUS_DISABLED)
+		return TEE_SUCCESS;
+	res = mbox_dt_register_chan_by_name(NULL, NULL, NULL, fdt, node,
+					    "psa_tfm",
+					    &handle_send_receive);
+	if (res) {
+		DMSG("No PSA mailbox channel %d", res);
+		return TEE_ERROR_DEFER_DRIVER_INIT;
+	}
+
+	prop = fdt_getprop(fdt, node, "shmem", &len);
+	if (!prop) {
+		EMSG("No Shared memory");
+		panic();
+	}
+
+	pnode = fdt_node_offset_by_phandle(fdt, fdt32_to_cpu(prop[0]));
+	if (pnode > 0) {
+		pa = fdt_reg_base_address(fdt, pnode);
+		shmem_size = fdt_reg_size(fdt, pnode);
+		shmem = core_mmu_add_mapping(MEM_AREA_IO_SEC, pa, shmem_size);
+	}
+	if (!shmem) {
+		EMSG("No Shared memory");
+		panic();
+	}
+
+	rse_register_cb(rse_mbox_send_data, rse_mbox_rcv_data,
+			rse_mbox_size_data);
+	/*  set a default client id */
+	rse_set_client_id(ST_PSA_CLIENT_ID);
+
+	return res;
+}
+service_init(stm32mp_start_psa_service);
diff --git a/core/arch/arm/plat-stm32mp2/psa_service/sub.mk b/core/arch/arm/plat-stm32mp2/psa_service/sub.mk
new file mode 100644
index 000000000..bab23b82b
--- /dev/null
+++ b/core/arch/arm/plat-stm32mp2/psa_service/sub.mk
@@ -0,0 +1,3 @@
+incdirs-y += .
+srcs-y += psa_client.c
+srcs-y += tfm_crypto.c
diff --git a/core/arch/arm/plat-stm32mp2/psa_service/tfm_crypto.c b/core/arch/arm/plat-stm32mp2/psa_service/tfm_crypto.c
new file mode 100644
index 000000000..f97f14e0f
--- /dev/null
+++ b/core/arch/arm/plat-stm32mp2/psa_service/tfm_crypto.c
@@ -0,0 +1,1641 @@
+// SPDX-License-Identifier: BSD-3-Clause
+/*
+ * Copyright (c) 2018-2023, Arm Limited. All rights reserved.
+ * Copyright (C) 2025, STMicroelectronics
+ *
+ */
+
+#include <stdlib.h>
+#include <string.h>
+#define MBEDTLS_PSA_CRYPTO_CLIENT
+#include "tfm_crypto_defs.h"
+
+#include "psa/client.h"
+#include "psa_manifest/sid.h"
+
+#define API_DISPATCH(in_vec, out_vec)          \
+	psa_call(RSE_CRYPTO_HANDLE, PSA_IPC_CALL,  \
+		 in_vec, IOVEC_LEN(in_vec),        \
+		 out_vec, IOVEC_LEN(out_vec))
+#define API_DISPATCH_NO_OUTVEC(in_vec)         \
+	psa_call(RSE_CRYPTO_HANDLE, PSA_IPC_CALL,  \
+		 in_vec, IOVEC_LEN(in_vec),        \
+		 (psa_outvec *)NULL, 0)
+
+psa_status_t psa_crypto_init(void)
+{
+	/* Service init is performed during TFM boot up,
+	 * so application level initialisation is empty
+	 */
+	return PSA_SUCCESS;
+}
+
+psa_status_t psa_open_key(psa_key_id_t id,
+			  psa_key_id_t *key)
+{
+	const struct tfm_crypto_pack_iovec iov = {
+		.function_id = TFM_CRYPTO_OPEN_KEY_SID,
+		.key_id = id,
+	};
+	psa_invec in_vec[] = {
+		{.base = &iov, .len = sizeof(struct tfm_crypto_pack_iovec)},
+	};
+	psa_outvec out_vec[] = {
+		{.base = key, .len = sizeof(psa_key_id_t)},
+	};
+
+	return API_DISPATCH(in_vec, out_vec);
+}
+
+psa_status_t psa_close_key(psa_key_id_t key)
+{
+	const struct tfm_crypto_pack_iovec iov = {
+		.function_id = TFM_CRYPTO_CLOSE_KEY_SID,
+		.key_id = key,
+	};
+	psa_invec in_vec[] = {
+		{.base = &iov, .len = sizeof(struct tfm_crypto_pack_iovec)},
+	};
+
+	return API_DISPATCH_NO_OUTVEC(in_vec);
+}
+
+psa_status_t psa_import_key(const psa_key_attributes_t *attributes,
+			    const uint8_t *data,
+			    size_t data_length,
+			    psa_key_id_t *key)
+{
+	struct tfm_crypto_pack_iovec iov = {
+		.function_id = TFM_CRYPTO_IMPORT_KEY_SID,
+	};
+	psa_invec in_vec[] = {
+		{.base = &iov, .len = sizeof(struct tfm_crypto_pack_iovec)},
+		{.base = attributes, .len = sizeof(psa_key_attributes_t)},
+		{.base = data, .len = data_length}
+	};
+	psa_outvec out_vec[] = {
+		{.base = key, .len = sizeof(psa_key_id_t)}
+	};
+
+	return API_DISPATCH(in_vec, out_vec);
+}
+
+psa_status_t psa_destroy_key(psa_key_id_t key)
+{
+	struct tfm_crypto_pack_iovec iov = {
+		.function_id = TFM_CRYPTO_DESTROY_KEY_SID,
+		.key_id = key,
+	};
+	psa_invec in_vec[] = {
+		{.base = &iov, .len = sizeof(struct tfm_crypto_pack_iovec)},
+	};
+
+	return API_DISPATCH_NO_OUTVEC(in_vec);
+}
+
+psa_status_t psa_get_key_attributes(psa_key_id_t key,
+				    psa_key_attributes_t *attributes)
+{
+	struct tfm_crypto_pack_iovec iov = {
+		.function_id = TFM_CRYPTO_GET_KEY_ATTRIBUTES_SID,
+		.key_id = key,
+	};
+	psa_invec in_vec[] = {
+		{.base = &iov, .len = sizeof(struct tfm_crypto_pack_iovec)},
+	};
+	psa_outvec out_vec[] = {
+		{.base = attributes, .len = sizeof(psa_key_attributes_t)},
+	};
+
+	return API_DISPATCH(in_vec, out_vec);
+}
+
+psa_status_t psa_export_key(psa_key_id_t key,
+			    uint8_t *data,
+			    size_t data_size,
+			    size_t *data_length)
+{
+	psa_status_t status;
+	struct tfm_crypto_pack_iovec iov = {
+		.function_id = TFM_CRYPTO_EXPORT_KEY_SID,
+		.key_id = key,
+	};
+	psa_invec in_vec[] = {
+		{.base = &iov, .len = sizeof(struct tfm_crypto_pack_iovec)},
+	};
+	psa_outvec out_vec[] = {
+		{.base = data, .len = data_size}
+	};
+
+	status = API_DISPATCH(in_vec, out_vec);
+
+	*data_length = out_vec[0].len;
+
+	return status;
+}
+
+psa_status_t psa_export_public_key(psa_key_id_t key,
+				   uint8_t *data,
+				   size_t data_size,
+				   size_t *data_length)
+{
+	psa_status_t status;
+	struct tfm_crypto_pack_iovec iov = {
+		.function_id = TFM_CRYPTO_EXPORT_PUBLIC_KEY_SID,
+		.key_id = key,
+	};
+
+	psa_invec in_vec[] = {
+		{.base = &iov, .len = sizeof(struct tfm_crypto_pack_iovec)},
+	};
+	psa_outvec out_vec[] = {
+		{.base = data, .len = data_size}
+	};
+
+	status = API_DISPATCH(in_vec, out_vec);
+
+	*data_length = out_vec[0].len;
+
+	return status;
+}
+
+psa_status_t psa_purge_key(psa_key_id_t key)
+{
+	struct tfm_crypto_pack_iovec iov = {
+		.function_id = TFM_CRYPTO_PURGE_KEY_SID,
+		.key_id = key,
+	};
+	psa_invec in_vec[] = {
+		{.base = &iov, .len = sizeof(struct tfm_crypto_pack_iovec)},
+	};
+
+	return API_DISPATCH_NO_OUTVEC(in_vec);
+}
+
+psa_status_t psa_copy_key(psa_key_id_t source_key,
+			  const psa_key_attributes_t *attributes,
+			  psa_key_id_t *target_key)
+{
+	struct tfm_crypto_pack_iovec iov = {
+		.function_id = TFM_CRYPTO_COPY_KEY_SID,
+		.key_id = source_key,
+	};
+
+	psa_invec in_vec[] = {
+		{.base = &iov, .len = sizeof(struct tfm_crypto_pack_iovec)},
+		{.base = attributes, .len = sizeof(psa_key_attributes_t)},
+	};
+
+	psa_outvec out_vec[] = {
+		{.base = target_key, .len = sizeof(psa_key_id_t)},
+	};
+
+	return API_DISPATCH(in_vec, out_vec);
+}
+
+psa_status_t psa_cipher_generate_iv(psa_cipher_operation_t *operation,
+				    unsigned char *iv,
+				    size_t iv_size,
+				    size_t *iv_length)
+{
+	psa_status_t status;
+	struct tfm_crypto_pack_iovec iov = {
+		.function_id = TFM_CRYPTO_CIPHER_GENERATE_IV_SID,
+		.op_handle = operation->handle,
+	};
+
+	psa_invec in_vec[] = {
+		{.base = &iov, .len = sizeof(struct tfm_crypto_pack_iovec)},
+	};
+	psa_outvec out_vec[] = {
+		{.base = iv, .len = iv_size},
+	};
+
+	status = API_DISPATCH(in_vec, out_vec);
+
+	*iv_length = out_vec[0].len;
+
+	return status;
+}
+
+psa_status_t psa_cipher_set_iv(psa_cipher_operation_t *operation,
+			       const unsigned char *iv,
+			       size_t iv_length)
+{
+	struct tfm_crypto_pack_iovec iov = {
+		.function_id = TFM_CRYPTO_CIPHER_SET_IV_SID,
+		.op_handle = operation->handle,
+	};
+
+	psa_invec in_vec[] = {
+		{.base = &iov, .len = sizeof(struct tfm_crypto_pack_iovec)},
+		{.base = iv, .len = iv_length},
+	};
+
+	return API_DISPATCH_NO_OUTVEC(in_vec);
+}
+
+psa_status_t psa_cipher_encrypt_setup(psa_cipher_operation_t *operation,
+				      psa_key_id_t key,
+				      psa_algorithm_t alg)
+{
+	struct tfm_crypto_pack_iovec iov = {
+		.function_id = TFM_CRYPTO_CIPHER_ENCRYPT_SETUP_SID,
+		.key_id = key,
+		.alg = alg,
+		.op_handle = operation->handle,
+	};
+
+	psa_invec in_vec[] = {
+		{.base = &iov, .len = sizeof(struct tfm_crypto_pack_iovec)},
+	};
+	psa_outvec out_vec[] = {
+		{.base = &(operation->handle), .len = sizeof(uint32_t)},
+	};
+
+	return API_DISPATCH(in_vec, out_vec);
+}
+
+psa_status_t psa_cipher_decrypt_setup(psa_cipher_operation_t *operation,
+				      psa_key_id_t key,
+				      psa_algorithm_t alg)
+{
+	struct tfm_crypto_pack_iovec iov = {
+		.function_id = TFM_CRYPTO_CIPHER_DECRYPT_SETUP_SID,
+		.key_id = key,
+		.alg = alg,
+		.op_handle = operation->handle,
+	};
+
+	psa_invec in_vec[] = {
+		{.base = &iov, .len = sizeof(struct tfm_crypto_pack_iovec)},
+	};
+	psa_outvec out_vec[] = {
+		{.base = &(operation->handle), .len = sizeof(uint32_t)},
+	};
+
+	return API_DISPATCH(in_vec, out_vec);
+}
+
+psa_status_t psa_cipher_update(psa_cipher_operation_t *operation,
+			       const uint8_t *input,
+			       size_t input_length,
+			       unsigned char *output,
+			       size_t output_size,
+			       size_t *output_length)
+{
+	psa_status_t status;
+	struct tfm_crypto_pack_iovec iov = {
+		.function_id = TFM_CRYPTO_CIPHER_UPDATE_SID,
+		.op_handle = operation->handle,
+	};
+
+	psa_invec in_vec[] = {
+		{.base = &iov, .len = sizeof(struct tfm_crypto_pack_iovec)},
+		{.base = input, .len = input_length},
+	};
+	psa_outvec out_vec[] = {
+		{.base = output, .len = output_size}
+	};
+
+	status = API_DISPATCH(in_vec, out_vec);
+
+	*output_length = out_vec[0].len;
+
+	return status;
+}
+
+psa_status_t psa_cipher_abort(psa_cipher_operation_t *operation)
+{
+	struct tfm_crypto_pack_iovec iov = {
+		.function_id = TFM_CRYPTO_CIPHER_ABORT_SID,
+		.op_handle = operation->handle,
+	};
+
+	psa_invec in_vec[] = {
+		{.base = &iov, .len = sizeof(struct tfm_crypto_pack_iovec)},
+	};
+	psa_outvec out_vec[] = {
+		{.base = &(operation->handle), .len = sizeof(uint32_t)},
+	};
+
+	return API_DISPATCH(in_vec, out_vec);
+}
+
+psa_status_t psa_cipher_finish(psa_cipher_operation_t *operation,
+			       uint8_t *output,
+			       size_t output_size,
+			       size_t *output_length)
+{
+	psa_status_t status;
+	struct tfm_crypto_pack_iovec iov = {
+		.function_id = TFM_CRYPTO_CIPHER_FINISH_SID,
+		.op_handle = operation->handle,
+	};
+
+	psa_invec in_vec[] = {
+		{.base = &iov, .len = sizeof(struct tfm_crypto_pack_iovec)},
+	};
+	psa_outvec out_vec[] = {
+		{.base = &(operation->handle), .len = sizeof(uint32_t)},
+		{.base = output, .len = output_size},
+	};
+
+	status = API_DISPATCH(in_vec, out_vec);
+
+	*output_length = out_vec[1].len;
+
+	return status;
+}
+
+psa_status_t psa_hash_setup(psa_hash_operation_t *operation,
+			    psa_algorithm_t alg)
+{
+	struct tfm_crypto_pack_iovec iov = {
+		.function_id = TFM_CRYPTO_HASH_SETUP_SID,
+		.alg = alg,
+		.op_handle = operation->handle,
+	};
+
+	psa_invec in_vec[] = {
+		{.base = &iov, .len = sizeof(struct tfm_crypto_pack_iovec)},
+	};
+	psa_outvec out_vec[] = {
+		{.base = &(operation->handle), .len = sizeof(uint32_t)},
+	};
+
+	return API_DISPATCH(in_vec, out_vec);
+}
+
+psa_status_t psa_hash_update(psa_hash_operation_t *operation,
+			     const uint8_t *input,
+			     size_t input_length)
+{
+	struct tfm_crypto_pack_iovec iov = {
+		.function_id = TFM_CRYPTO_HASH_UPDATE_SID,
+		.op_handle = operation->handle,
+	};
+
+	psa_invec in_vec[] = {
+		{.base = &iov, .len = sizeof(struct tfm_crypto_pack_iovec)},
+		{.base = input, .len = input_length},
+	};
+
+	return API_DISPATCH_NO_OUTVEC(in_vec);
+}
+
+psa_status_t psa_hash_finish(psa_hash_operation_t *operation,
+			     uint8_t *hash,
+			     size_t hash_size,
+			     size_t *hash_length)
+{
+	psa_status_t status;
+	struct tfm_crypto_pack_iovec iov = {
+		.function_id = TFM_CRYPTO_HASH_FINISH_SID,
+		.op_handle = operation->handle,
+	};
+
+	psa_invec in_vec[] = {
+		{.base = &iov, .len = sizeof(struct tfm_crypto_pack_iovec)},
+	};
+	psa_outvec out_vec[] = {
+		{.base = &(operation->handle), .len = sizeof(uint32_t)},
+		{.base = hash, .len = hash_size},
+	};
+
+	status = API_DISPATCH(in_vec, out_vec);
+
+	*hash_length = out_vec[1].len;
+
+	return status;
+}
+
+psa_status_t psa_hash_verify(psa_hash_operation_t *operation,
+			     const uint8_t *hash,
+			     size_t hash_length)
+{
+	struct tfm_crypto_pack_iovec iov = {
+		.function_id = TFM_CRYPTO_HASH_VERIFY_SID,
+		.op_handle = operation->handle,
+	};
+
+	psa_invec in_vec[] = {
+		{.base = &iov, .len = sizeof(struct tfm_crypto_pack_iovec)},
+		{.base = hash, .len = hash_length},
+	};
+	psa_outvec out_vec[] = {
+		{.base = &(operation->handle), .len = sizeof(uint32_t)},
+	};
+
+	return API_DISPATCH(in_vec, out_vec);
+}
+
+psa_status_t psa_hash_abort(psa_hash_operation_t *operation)
+{
+	struct tfm_crypto_pack_iovec iov = {
+		.function_id = TFM_CRYPTO_HASH_ABORT_SID,
+		.op_handle = operation->handle,
+	};
+
+	psa_invec in_vec[] = {
+		{.base = &iov, .len = sizeof(struct tfm_crypto_pack_iovec)},
+	};
+	psa_outvec out_vec[] = {
+		{.base = &(operation->handle), .len = sizeof(uint32_t)},
+	};
+
+	return API_DISPATCH(in_vec, out_vec);
+}
+
+psa_status_t psa_hash_clone(const psa_hash_operation_t *source_operation,
+			    psa_hash_operation_t *target_operation)
+{
+	struct tfm_crypto_pack_iovec iov = {
+		.function_id = TFM_CRYPTO_HASH_CLONE_SID,
+		.op_handle = source_operation->handle,
+	};
+
+	psa_invec in_vec[] = {
+		{.base = &iov, .len = sizeof(struct tfm_crypto_pack_iovec)},
+		{.base = &(target_operation->handle),
+			.len = sizeof(target_operation->handle)},
+	};
+	psa_outvec out_vec[] = {
+		{.base = &(target_operation->handle),
+			.len = sizeof(target_operation->handle)},
+	};
+
+	if (target_operation && (target_operation->handle != 0))
+		return PSA_ERROR_BAD_STATE;
+
+	return API_DISPATCH(in_vec, out_vec);
+}
+
+psa_status_t psa_hash_compute(psa_algorithm_t alg,
+			      const uint8_t *input,
+			      size_t input_length,
+			      uint8_t *hash,
+			      size_t hash_size,
+			      size_t *hash_length)
+{
+	psa_status_t status;
+	struct tfm_crypto_pack_iovec iov = {
+		.function_id = TFM_CRYPTO_HASH_COMPUTE_SID,
+		.alg = alg,
+	};
+
+	psa_invec in_vec[2] = {
+		{.base = &iov, .len = sizeof(struct tfm_crypto_pack_iovec)},
+		{.base = input, .len = input_length},
+	};
+
+	psa_outvec out_vec[1] = {
+		{.base = hash, .len = hash_size}
+	};
+
+	status = API_DISPATCH(in_vec, out_vec);
+
+	*hash_length = out_vec[0].len;
+
+	return status;
+}
+
+psa_status_t psa_hash_compare(psa_algorithm_t alg,
+			      const uint8_t *input,
+			      size_t input_length,
+			      const uint8_t *hash,
+			      size_t hash_length)
+{
+	struct tfm_crypto_pack_iovec iov = {
+		.function_id = TFM_CRYPTO_HASH_COMPARE_SID,
+		.alg = alg,
+	};
+
+	psa_invec in_vec[] = {
+		{.base = &iov, .len = sizeof(struct tfm_crypto_pack_iovec)},
+		{.base = input, .len = input_length},
+		{.base = hash, .len = hash_length},
+	};
+
+	return API_DISPATCH_NO_OUTVEC(in_vec);
+}
+
+psa_status_t psa_mac_sign_setup(psa_mac_operation_t *operation,
+				psa_key_id_t key,
+				psa_algorithm_t alg)
+{
+	struct tfm_crypto_pack_iovec iov = {
+		.function_id = TFM_CRYPTO_MAC_SIGN_SETUP_SID,
+		.key_id = key,
+		.alg = alg,
+		.op_handle = operation->handle,
+	};
+
+	psa_invec in_vec[] = {
+		{.base = &iov, .len = sizeof(struct tfm_crypto_pack_iovec)},
+	};
+	psa_outvec out_vec[] = {
+		{.base = &(operation->handle), .len = sizeof(uint32_t)},
+	};
+
+	return API_DISPATCH(in_vec, out_vec);
+}
+
+psa_status_t psa_mac_verify_setup(psa_mac_operation_t *operation,
+				  psa_key_id_t key,
+				  psa_algorithm_t alg)
+{
+	struct tfm_crypto_pack_iovec iov = {
+		.function_id = TFM_CRYPTO_MAC_VERIFY_SETUP_SID,
+		.key_id = key,
+		.alg = alg,
+		.op_handle = operation->handle,
+	};
+
+	psa_invec in_vec[] = {
+		{.base = &iov, .len = sizeof(struct tfm_crypto_pack_iovec)},
+	};
+	psa_outvec out_vec[] = {
+		{.base = &(operation->handle), .len = sizeof(uint32_t)},
+	};
+
+	return API_DISPATCH(in_vec, out_vec);
+}
+
+psa_status_t psa_mac_update(psa_mac_operation_t *operation,
+			    const uint8_t *input,
+			    size_t input_length)
+{
+	struct tfm_crypto_pack_iovec iov = {
+		.function_id = TFM_CRYPTO_MAC_UPDATE_SID,
+		.op_handle = operation->handle,
+	};
+
+	psa_invec in_vec[] = {
+		{.base = &iov, .len = sizeof(struct tfm_crypto_pack_iovec)},
+		{.base = input, .len = input_length},
+	};
+
+	return API_DISPATCH_NO_OUTVEC(in_vec);
+}
+
+psa_status_t psa_mac_sign_finish(psa_mac_operation_t *operation,
+				 uint8_t *mac,
+				 size_t mac_size,
+				 size_t *mac_length)
+{
+	psa_status_t status;
+	struct tfm_crypto_pack_iovec iov = {
+		.function_id = TFM_CRYPTO_MAC_SIGN_FINISH_SID,
+		.op_handle = operation->handle,
+	};
+
+	psa_invec in_vec[] = {
+		{.base = &iov, .len = sizeof(struct tfm_crypto_pack_iovec)},
+	};
+	psa_outvec out_vec[] = {
+		{.base = &(operation->handle), .len = sizeof(uint32_t)},
+		{.base = mac, .len = mac_size},
+	};
+
+	status = API_DISPATCH(in_vec, out_vec);
+
+	*mac_length = out_vec[1].len;
+
+	return status;
+}
+
+psa_status_t psa_mac_verify_finish(psa_mac_operation_t *operation,
+				   const uint8_t *mac,
+				   size_t mac_length)
+{
+	struct tfm_crypto_pack_iovec iov = {
+		.function_id = TFM_CRYPTO_MAC_VERIFY_FINISH_SID,
+		.op_handle = operation->handle,
+	};
+
+	psa_invec in_vec[] = {
+		{.base = &iov, .len = sizeof(struct tfm_crypto_pack_iovec)},
+		{.base = mac, .len = mac_length},
+	};
+	psa_outvec out_vec[] = {
+		{.base = &(operation->handle), .len = sizeof(uint32_t)},
+	};
+
+	return API_DISPATCH(in_vec, out_vec);
+}
+
+psa_status_t psa_mac_abort(psa_mac_operation_t *operation)
+{
+	struct tfm_crypto_pack_iovec iov = {
+		.function_id = TFM_CRYPTO_MAC_ABORT_SID,
+		.op_handle = operation->handle,
+	};
+
+	psa_invec in_vec[] = {
+		{.base = &iov, .len = sizeof(struct tfm_crypto_pack_iovec)},
+	};
+	psa_outvec out_vec[] = {
+		{.base = &(operation->handle), .len = sizeof(uint32_t)},
+	};
+
+	return API_DISPATCH(in_vec, out_vec);
+}
+
+psa_status_t psa_aead_encrypt(psa_key_id_t key,
+			      psa_algorithm_t alg,
+			      const uint8_t *nonce,
+			      size_t nonce_length,
+			      const uint8_t *additional_data,
+			      size_t additional_data_length,
+			      const uint8_t *plaintext,
+			      size_t plaintext_length,
+			      uint8_t *ciphertext,
+			      size_t ciphertext_size,
+			      size_t *ciphertext_length)
+{
+	psa_status_t status;
+	struct tfm_crypto_pack_iovec iov = {
+		.function_id = TFM_CRYPTO_AEAD_ENCRYPT_SID,
+		.key_id = key,
+		.alg = alg,
+		.aead_in = {.nonce = {0}, .nonce_length = 0}
+	};
+	psa_invec in_vec[] = {
+		{.base = NULL, .len = 0},
+		{.base = plaintext, .len = plaintext_length},
+		{.base = additional_data, .len = additional_data_length},
+	};
+	psa_outvec out_vec[] = {
+		{.base = ciphertext, .len = ciphertext_size},
+	};
+	size_t in_len = IOVEC_LEN(in_vec);
+
+
+	/* Sanitize the optional input */
+	if ((!additional_data) && (additional_data_length))
+		return PSA_ERROR_INVALID_ARGUMENT;
+
+	if (nonce_length > TFM_CRYPTO_MAX_NONCE_LENGTH)
+		return PSA_ERROR_INVALID_ARGUMENT;
+
+	if (nonce) {
+		for (size_t idx = 0; idx < nonce_length; idx++)
+			iov.aead_in.nonce[idx] = nonce[idx];
+		iov.aead_in.nonce_length = nonce_length;
+	}
+
+	in_vec[0].base = &iov;
+	in_vec[0].len = sizeof(struct tfm_crypto_pack_iovec);
+
+	if (!additional_data)
+		in_len--;
+	status = psa_call(RSE_CRYPTO_HANDLE, PSA_IPC_CALL, in_vec, in_len,
+			  out_vec, IOVEC_LEN(out_vec));
+
+	*ciphertext_length = out_vec[0].len;
+
+	return status;
+}
+
+psa_status_t psa_aead_decrypt(psa_key_id_t key,
+			      psa_algorithm_t alg,
+			      const uint8_t *nonce,
+			      size_t nonce_length,
+			      const uint8_t *additional_data,
+			      size_t additional_data_length,
+			      const uint8_t *ciphertext,
+			      size_t ciphertext_length,
+			      uint8_t *plaintext,
+			      size_t plaintext_size,
+			      size_t *plaintext_length)
+{
+	psa_status_t status;
+	struct tfm_crypto_pack_iovec iov = {
+		.function_id = TFM_CRYPTO_AEAD_DECRYPT_SID,
+		.key_id = key,
+		.alg = alg,
+		.aead_in = {.nonce = {0}, .nonce_length = 0}
+	};
+
+	psa_invec in_vec[] = {
+		{.base = NULL, .len = 0},
+		{.base = ciphertext, .len = ciphertext_length},
+		{.base = additional_data, .len = additional_data_length},
+	};
+	psa_outvec out_vec[] = {
+		{.base = plaintext, .len = plaintext_size},
+	};
+	size_t in_len = IOVEC_LEN(in_vec);
+
+
+	/* Sanitize the optional input */
+	if ((!additional_data) && (additional_data_length != 0))
+		return PSA_ERROR_INVALID_ARGUMENT;
+
+	if (nonce_length > TFM_CRYPTO_MAX_NONCE_LENGTH)
+		return PSA_ERROR_INVALID_ARGUMENT;
+
+	if (nonce) {
+		for (size_t idx = 0; idx < nonce_length; idx++)
+			iov.aead_in.nonce[idx] = nonce[idx];
+		iov.aead_in.nonce_length = nonce_length;
+	}
+
+	in_vec[0].base = &iov;
+	in_vec[0].len = sizeof(struct tfm_crypto_pack_iovec);
+
+	if (!additional_data)
+		in_len--;
+	status = psa_call(RSE_CRYPTO_HANDLE, PSA_IPC_CALL, in_vec, in_len,
+			  out_vec, IOVEC_LEN(out_vec));
+
+	*plaintext_length = out_vec[0].len;
+
+	return status;
+}
+
+psa_status_t psa_aead_encrypt_setup(psa_aead_operation_t *operation,
+				    psa_key_id_t key,
+				    psa_algorithm_t alg)
+{
+	psa_status_t status;
+	struct tfm_crypto_pack_iovec iov = {
+		.function_id = TFM_CRYPTO_AEAD_ENCRYPT_SETUP_SID,
+		.key_id = key,
+		.alg = alg,
+		.op_handle = operation->handle,
+	};
+
+	psa_invec in_vec[] = {
+		{.base = &iov, .len = sizeof(struct tfm_crypto_pack_iovec)}
+	};
+	psa_outvec out_vec[] = {
+		{.base = &(operation->handle), .len = sizeof(uint32_t)}
+	};
+
+	status = API_DISPATCH(in_vec, out_vec);
+	return status;
+}
+
+psa_status_t psa_aead_decrypt_setup(psa_aead_operation_t *operation,
+				    psa_key_id_t key,
+				    psa_algorithm_t alg)
+{
+	psa_status_t status;
+	struct tfm_crypto_pack_iovec iov = {
+		.function_id = TFM_CRYPTO_AEAD_DECRYPT_SETUP_SID,
+		.key_id = key,
+		.alg = alg,
+		.op_handle = operation->handle,
+	};
+
+	psa_invec in_vec[] = {
+		{.base = &iov, .len = sizeof(struct tfm_crypto_pack_iovec)}
+	};
+	psa_outvec out_vec[] = {
+		{.base = &(operation->handle), .len = sizeof(uint32_t)}
+	};
+
+	status = API_DISPATCH(in_vec, out_vec);
+	return status;
+}
+
+psa_status_t psa_aead_generate_nonce(psa_aead_operation_t *operation,
+				     uint8_t *nonce,
+				     size_t nonce_size,
+				     size_t *nonce_length)
+{
+	psa_status_t status;
+	struct tfm_crypto_pack_iovec iov = {
+		.function_id = TFM_CRYPTO_AEAD_GENERATE_NONCE_SID,
+		.op_handle = operation->handle,
+	};
+
+	psa_invec in_vec[] = {
+		{.base = &iov, .len = sizeof(struct tfm_crypto_pack_iovec)},
+	};
+	psa_outvec out_vec[] = {
+		{.base = nonce, .len = nonce_size}
+	};
+
+	status = API_DISPATCH(in_vec, out_vec);
+
+	*nonce_length = out_vec[0].len;
+	return status;
+}
+
+psa_status_t psa_aead_set_nonce(psa_aead_operation_t *operation,
+				const uint8_t *nonce,
+				size_t nonce_length)
+{
+	psa_status_t status;
+	struct tfm_crypto_pack_iovec iov = {
+		.function_id = TFM_CRYPTO_AEAD_SET_NONCE_SID,
+		.op_handle = operation->handle,
+	};
+
+	psa_invec in_vec[] = {
+		{.base = &iov, .len = sizeof(struct tfm_crypto_pack_iovec)},
+		{.base = nonce, .len = nonce_length}
+	};
+
+	status = API_DISPATCH_NO_OUTVEC(in_vec);
+	return status;
+}
+
+psa_status_t psa_aead_set_lengths(psa_aead_operation_t *operation,
+				  size_t ad_length,
+				  size_t plaintext_length)
+{
+	psa_status_t status;
+	struct tfm_crypto_pack_iovec iov = {
+		.function_id = TFM_CRYPTO_AEAD_SET_LENGTHS_SID,
+		.ad_length = ad_length,
+		.plaintext_length = plaintext_length,
+		.op_handle = operation->handle,
+	};
+
+	psa_invec in_vec[] = {
+		{.base = &iov, .len = sizeof(struct tfm_crypto_pack_iovec)},
+	};
+
+	status = API_DISPATCH_NO_OUTVEC(in_vec);
+	return status;
+}
+
+psa_status_t psa_aead_update_ad(psa_aead_operation_t *operation,
+				const uint8_t *input,
+				size_t input_length)
+{
+	psa_status_t status;
+	struct tfm_crypto_pack_iovec iov = {
+		.function_id = TFM_CRYPTO_AEAD_UPDATE_AD_SID,
+		.op_handle = operation->handle,
+	};
+	psa_invec in_vec[] = {
+		{.base = &iov, .len = sizeof(struct tfm_crypto_pack_iovec)},
+		{.base = input, .len = input_length}
+	};
+
+	size_t in_len = IOVEC_LEN(in_vec);
+
+
+	/* Sanitize the optional input */
+	if ((!input) && (input_length))
+		return PSA_ERROR_INVALID_ARGUMENT;
+
+	if (!input)
+		in_len--;
+
+	status = psa_call(RSE_CRYPTO_HANDLE, PSA_IPC_CALL, in_vec, in_len,
+			  NULL, 0);
+	return status;
+}
+
+psa_status_t psa_aead_update(psa_aead_operation_t *operation,
+			     const uint8_t *input,
+			     size_t input_length,
+			     uint8_t *output,
+			     size_t output_size,
+			     size_t *output_length)
+{
+	psa_status_t status;
+	struct tfm_crypto_pack_iovec iov = {
+		.function_id = TFM_CRYPTO_AEAD_UPDATE_SID,
+		.op_handle = operation->handle,
+	};
+	psa_invec in_vec[] = {
+		{.base = &iov, .len = sizeof(struct tfm_crypto_pack_iovec)},
+		{.base = input, .len = input_length}
+	};
+	psa_outvec out_vec[] = {
+		{.base = output, .len = output_size},
+	};
+
+	size_t in_len = IOVEC_LEN(in_vec);
+
+	/* Sanitize the optional input */
+	if ((!input) && (input_length))
+		return PSA_ERROR_INVALID_ARGUMENT;
+
+	if (input == NULL)
+		in_len--;
+
+	status = psa_call(RSE_CRYPTO_HANDLE, PSA_IPC_CALL, in_vec, in_len,
+			  out_vec, IOVEC_LEN(out_vec));
+
+	*output_length = out_vec[0].len;
+	return status;
+}
+
+psa_status_t psa_aead_finish(psa_aead_operation_t *operation,
+			     uint8_t *ciphertext,
+			     size_t ciphertext_size,
+			     size_t *ciphertext_length,
+			     uint8_t *tag,
+			     size_t tag_size,
+			     size_t *tag_length)
+{
+	psa_status_t status;
+	struct tfm_crypto_pack_iovec iov = {
+		.function_id = TFM_CRYPTO_AEAD_FINISH_SID,
+		.op_handle = operation->handle,
+	};
+
+	psa_invec in_vec[1] = {
+		{.base = &iov, .len = sizeof(struct tfm_crypto_pack_iovec)},
+	};
+	psa_outvec out_vec[3] = {
+		{.base = &(operation->handle), .len = sizeof(uint32_t)},
+		{.base = tag, .len = tag_size},
+		{.base = ciphertext, .len = ciphertext_size}
+	};
+
+	size_t out_len = IOVEC_LEN(out_vec);
+
+
+	/* Sanitize the optional output */
+	if ((ciphertext == NULL) && (ciphertext_size != 0))
+		return PSA_ERROR_INVALID_ARGUMENT;
+
+	if (ciphertext == NULL || ciphertext_size == 0)
+		out_len--;
+
+	if ((out_len == 3) && (ciphertext_length == NULL))
+		return PSA_ERROR_INVALID_ARGUMENT;
+
+	status = psa_call(RSE_CRYPTO_HANDLE, PSA_IPC_CALL,
+			  in_vec, IOVEC_LEN(in_vec),
+			  out_vec, out_len);
+
+	if (out_len == 3)
+		*ciphertext_length = out_vec[2].len;
+	else
+		*ciphertext_length = 0;
+
+	*tag_length = out_vec[1].len;
+
+	return status;
+}
+
+psa_status_t psa_aead_verify(psa_aead_operation_t *operation,
+			     uint8_t *plaintext,
+			     size_t plaintext_size,
+			     size_t *plaintext_length,
+			     const uint8_t *tag,
+			     size_t tag_length)
+{
+	psa_status_t status;
+	struct tfm_crypto_pack_iovec iov = {
+		.function_id = TFM_CRYPTO_AEAD_VERIFY_SID,
+		.op_handle = operation->handle,
+	};
+
+	psa_invec in_vec[] = {
+		{.base = &iov, .len = sizeof(struct tfm_crypto_pack_iovec)},
+		{.base = tag, .len = tag_length}
+	};
+	psa_outvec out_vec[] = {
+		{.base = &(operation->handle), .len = sizeof(uint32_t)},
+		{.base = plaintext, .len = plaintext_size}
+	};
+
+	size_t out_len = IOVEC_LEN(out_vec);
+
+	/* Sanitize the optional output */
+	if ((!plaintext) && (plaintext_size))
+		return PSA_ERROR_INVALID_ARGUMENT;
+
+	if (!plaintext || plaintext_size == 0)
+		out_len--;
+	if ((out_len == 2) && (!plaintext_length))
+		return PSA_ERROR_INVALID_ARGUMENT;
+
+	status = psa_call(RSE_CRYPTO_HANDLE, PSA_IPC_CALL,
+			  in_vec, IOVEC_LEN(in_vec),
+			  out_vec, out_len);
+
+	if (out_len == 2)
+		*plaintext_length = out_vec[1].len;
+	else
+		*plaintext_length = 0;
+
+	return status;
+}
+
+psa_status_t psa_aead_abort(psa_aead_operation_t *operation)
+{
+	struct tfm_crypto_pack_iovec iov = {
+		.function_id = TFM_CRYPTO_AEAD_ABORT_SID,
+		.op_handle = operation->handle,
+	};
+
+	psa_invec in_vec[] = {
+		{.base = &iov, .len = sizeof(struct tfm_crypto_pack_iovec)},
+	};
+	psa_outvec out_vec[] = {
+		{.base = &(operation->handle), .len = sizeof(uint32_t)},
+	};
+
+	return API_DISPATCH(in_vec, out_vec);
+}
+
+psa_status_t psa_sign_message(psa_key_id_t key,
+			      psa_algorithm_t alg,
+			      const uint8_t *input,
+			      size_t input_length,
+			      uint8_t *signature,
+			      size_t signature_size,
+			      size_t *signature_length)
+{
+	psa_status_t status;
+	struct tfm_crypto_pack_iovec iov = {
+		.function_id = TFM_CRYPTO_ASYMMETRIC_SIGN_MESSAGE_SID,
+		.key_id = key,
+		.alg = alg,
+	};
+
+	psa_invec in_vec[] = {
+		{.base = &iov, .len = sizeof(struct tfm_crypto_pack_iovec)},
+		{.base = input, .len = input_length},
+	};
+	psa_outvec out_vec[] = {
+		{.base = signature, .len = signature_size},
+	};
+
+	status = API_DISPATCH(in_vec, out_vec);
+
+	*signature_length = out_vec[0].len;
+	return status;
+}
+
+psa_status_t psa_verify_message(psa_key_id_t key,
+				psa_algorithm_t alg,
+				const uint8_t *input,
+				size_t input_length,
+				const uint8_t *signature,
+				size_t signature_length)
+{
+	struct tfm_crypto_pack_iovec iov = {
+		.function_id = TFM_CRYPTO_ASYMMETRIC_VERIFY_MESSAGE_SID,
+		.key_id = key,
+		.alg = alg
+	};
+
+	psa_invec in_vec[] = {
+		{.base = &iov, .len = sizeof(struct tfm_crypto_pack_iovec)},
+		{.base = input, .len = input_length},
+		{.base = signature, .len = signature_length}
+	};
+
+	return API_DISPATCH_NO_OUTVEC(in_vec);
+}
+
+psa_status_t psa_sign_hash(psa_key_id_t key,
+			   psa_algorithm_t alg,
+			   const uint8_t *hash,
+			   size_t hash_length,
+			   uint8_t *signature,
+			   size_t signature_size,
+			   size_t *signature_length)
+{
+	psa_status_t status;
+	struct tfm_crypto_pack_iovec iov = {
+		.function_id = TFM_CRYPTO_ASYMMETRIC_SIGN_HASH_SID,
+		.key_id = key,
+		.alg = alg,
+	};
+
+	psa_invec in_vec[] = {
+		{.base = &iov, .len = sizeof(struct tfm_crypto_pack_iovec)},
+		{.base = hash, .len = hash_length},
+	};
+	psa_outvec out_vec[] = {
+		{.base = signature, .len = signature_size},
+	};
+
+	status = API_DISPATCH(in_vec, out_vec);
+
+	*signature_length = out_vec[0].len;
+
+	return status;
+}
+
+psa_status_t psa_verify_hash(psa_key_id_t key,
+			     psa_algorithm_t alg,
+			     const uint8_t *hash,
+			     size_t hash_length,
+			     const uint8_t *signature,
+			     size_t signature_length)
+{
+	struct tfm_crypto_pack_iovec iov = {
+		.function_id = TFM_CRYPTO_ASYMMETRIC_VERIFY_HASH_SID,
+		.key_id = key,
+		.alg = alg
+	};
+
+	psa_invec in_vec[] = {
+		{.base = &iov, .len = sizeof(struct tfm_crypto_pack_iovec)},
+		{.base = hash, .len = hash_length},
+		{.base = signature, .len = signature_length}
+	};
+
+	return API_DISPATCH_NO_OUTVEC(in_vec);
+}
+
+psa_status_t psa_asymmetric_encrypt(psa_key_id_t key,
+				    psa_algorithm_t alg,
+				    const uint8_t *input,
+				    size_t input_length,
+				    const uint8_t *salt,
+				    size_t salt_length,
+				    uint8_t *output,
+				    size_t output_size,
+				    size_t *output_length)
+{
+	psa_status_t status;
+	struct tfm_crypto_pack_iovec iov = {
+		.function_id = TFM_CRYPTO_ASYMMETRIC_ENCRYPT_SID,
+		.key_id = key,
+		.alg = alg
+	};
+
+	psa_invec in_vec[] = {
+		{.base = &iov, .len = sizeof(struct tfm_crypto_pack_iovec)},
+		{.base = input, .len = input_length},
+		{.base = salt, .len = salt_length}
+	};
+
+	psa_outvec out_vec[] = {
+		{.base = output, .len = output_size},
+	};
+
+	size_t in_len = IOVEC_LEN(in_vec);
+
+	/* Sanitize the optional input */
+	if ((!salt) && (salt_length))
+		return PSA_ERROR_INVALID_ARGUMENT;
+
+	if (!salt)
+		in_len--;
+
+	status = psa_call(RSE_CRYPTO_HANDLE, PSA_IPC_CALL, in_vec, in_len,
+			  out_vec, IOVEC_LEN(out_vec));
+
+	*output_length = out_vec[0].len;
+
+	return status;
+}
+
+psa_status_t psa_asymmetric_decrypt(psa_key_id_t key,
+				    psa_algorithm_t alg,
+				    const uint8_t *input,
+				    size_t input_length,
+				    const uint8_t *salt,
+				    size_t salt_length,
+				    uint8_t *output,
+				    size_t output_size,
+				    size_t *output_length)
+{
+	psa_status_t status;
+	struct tfm_crypto_pack_iovec iov = {
+		.function_id = TFM_CRYPTO_ASYMMETRIC_DECRYPT_SID,
+		.key_id = key,
+		.alg = alg
+	};
+
+	psa_invec in_vec[] = {
+		{.base = &iov, .len = sizeof(struct tfm_crypto_pack_iovec)},
+		{.base = input, .len = input_length},
+		{.base = salt, .len = salt_length}
+	};
+
+	psa_outvec out_vec[] = {
+		{.base = output, .len = output_size},
+	};
+
+	size_t in_len = IOVEC_LEN(in_vec);
+
+	/* Sanitize the optional input */
+	if ((!salt) && (salt_length))
+		return PSA_ERROR_INVALID_ARGUMENT;
+
+	if (!salt)
+		in_len--;
+	status = psa_call(RSE_CRYPTO_HANDLE, PSA_IPC_CALL, in_vec, in_len,
+			  out_vec, IOVEC_LEN(out_vec));
+
+	*output_length = out_vec[0].len;
+
+	return status;
+}
+
+psa_status_t psa_key_derivation_get_capacity(const
+					     psa_key_derivation_operation_t
+					     *operation,
+					     size_t *capacity)
+{
+	struct tfm_crypto_pack_iovec iov = {
+		.function_id = TFM_CRYPTO_KEY_DERIVATION_GET_CAPACITY_SID,
+		.op_handle = operation->handle,
+	};
+
+	psa_invec in_vec[] = {
+		{.base = &iov, .len = sizeof(struct tfm_crypto_pack_iovec)},
+	};
+
+	psa_outvec out_vec[] = {
+		{.base = capacity, .len = sizeof(size_t)},
+	};
+
+	return API_DISPATCH(in_vec, out_vec);
+}
+
+psa_status_t psa_key_derivation_output_bytes(psa_key_derivation_operation_t
+					     *operation,
+					     uint8_t *output,
+					     size_t output_length)
+{
+	struct tfm_crypto_pack_iovec iov = {
+		.function_id = TFM_CRYPTO_KEY_DERIVATION_OUTPUT_BYTES_SID,
+		.op_handle = operation->handle,
+	};
+
+	psa_invec in_vec[] = {
+		{.base = &iov, .len = sizeof(struct tfm_crypto_pack_iovec)},
+	};
+
+	psa_outvec out_vec[] = {
+		{.base = output, .len = output_length},
+	};
+
+	return API_DISPATCH(in_vec, out_vec);
+}
+
+psa_status_t psa_key_derivation_input_key(psa_key_derivation_operation_t
+					  *operation,
+					  psa_key_derivation_step_t step,
+					  psa_key_id_t key)
+{
+	struct tfm_crypto_pack_iovec iov = {
+		.function_id = TFM_CRYPTO_KEY_DERIVATION_INPUT_KEY_SID,
+		.key_id = key,
+		.step = step,
+		.op_handle = operation->handle,
+	};
+
+	psa_invec in_vec[] = {
+		{.base = &iov, .len = sizeof(struct tfm_crypto_pack_iovec)},
+	};
+
+	return API_DISPATCH_NO_OUTVEC(in_vec);
+}
+
+psa_status_t psa_key_derivation_abort(psa_key_derivation_operation_t *operation)
+{
+	struct tfm_crypto_pack_iovec iov = {
+		.function_id = TFM_CRYPTO_KEY_DERIVATION_ABORT_SID,
+		.op_handle = operation->handle,
+	};
+
+	psa_invec in_vec[] = {
+		{.base = &iov, .len = sizeof(struct tfm_crypto_pack_iovec)},
+	};
+
+	psa_outvec out_vec[] = {
+		{.base = &(operation->handle), .len = sizeof(uint32_t)},
+	};
+
+	return API_DISPATCH(in_vec, out_vec);
+}
+
+psa_status_t psa_key_derivation_key_agreement(psa_key_derivation_operation_t
+					      *operation,
+					      psa_key_derivation_step_t step,
+					      psa_key_id_t private_key,
+					      const uint8_t *peer_key,
+					      size_t peer_key_length)
+{
+	struct tfm_crypto_pack_iovec iov = {
+		.function_id = TFM_CRYPTO_KEY_DERIVATION_KEY_AGREEMENT_SID,
+		.key_id = private_key,
+		.step = step,
+		.op_handle = operation->handle,
+	};
+
+	psa_invec in_vec[] = {
+		{.base = &iov, .len = sizeof(struct tfm_crypto_pack_iovec)},
+		{.base = peer_key, .len = peer_key_length},
+	};
+
+	return API_DISPATCH_NO_OUTVEC(in_vec);
+}
+
+psa_status_t psa_generate_random(uint8_t *output,
+				 size_t output_size)
+{
+	struct tfm_crypto_pack_iovec iov = {
+		.function_id = TFM_CRYPTO_GENERATE_RANDOM_SID,
+	};
+
+	psa_invec in_vec[] = {
+		{.base = &iov, .len = sizeof(struct tfm_crypto_pack_iovec)},
+	};
+
+	psa_outvec out_vec[] = {
+		{.base = output, .len = output_size},
+	};
+
+	if (!output_size)
+		return PSA_SUCCESS;
+
+	return API_DISPATCH(in_vec, out_vec);
+}
+
+psa_status_t psa_generate_key(const psa_key_attributes_t *attributes,
+			      psa_key_id_t *key)
+{
+	struct tfm_crypto_pack_iovec iov = {
+		.function_id = TFM_CRYPTO_GENERATE_KEY_SID,
+	};
+
+	psa_invec in_vec[] = {
+		{.base = &iov, .len = sizeof(struct tfm_crypto_pack_iovec)},
+		{.base = attributes, .len = sizeof(psa_key_attributes_t)},
+	};
+
+	psa_outvec out_vec[] = {
+		{.base = key, .len = sizeof(psa_key_id_t)},
+	};
+
+	return API_DISPATCH(in_vec, out_vec);
+}
+
+psa_status_t psa_mac_compute(psa_key_id_t key,
+			     psa_algorithm_t alg,
+			     const uint8_t *input,
+			     size_t input_length,
+			     uint8_t *mac,
+			     size_t mac_size,
+			     size_t *mac_length)
+{
+	psa_status_t status;
+	struct tfm_crypto_pack_iovec iov = {
+		.function_id = TFM_CRYPTO_MAC_COMPUTE_SID,
+		.key_id = key,
+		.alg = alg,
+	};
+
+	psa_invec in_vec[] = {
+		{.base = &iov, .len = sizeof(struct tfm_crypto_pack_iovec)},
+		{.base = input, .len = input_length},
+	};
+	psa_outvec out_vec[] = {
+		{.base = mac, .len = mac_size},
+	};
+
+	status = API_DISPATCH(in_vec, out_vec);
+
+	*mac_length = out_vec[0].len;
+	return status;
+}
+
+psa_status_t psa_mac_verify(psa_key_id_t key,
+			    psa_algorithm_t alg,
+			    const uint8_t *input,
+			    size_t input_length,
+			    const uint8_t *mac,
+			    const size_t mac_length)
+{
+	struct tfm_crypto_pack_iovec iov = {
+		.function_id = TFM_CRYPTO_MAC_VERIFY_SID,
+		.key_id = key,
+		.alg = alg,
+	};
+
+	psa_invec in_vec[] = {
+		{.base = &iov, .len = sizeof(struct tfm_crypto_pack_iovec)},
+		{.base = input, .len = input_length},
+		{.base = mac, .len = mac_length},
+	};
+
+	return API_DISPATCH_NO_OUTVEC(in_vec);
+}
+
+psa_status_t psa_cipher_encrypt(psa_key_id_t key,
+				psa_algorithm_t alg,
+				const uint8_t *input,
+				size_t input_length,
+				uint8_t *output,
+				size_t output_size,
+				size_t *output_length)
+{
+	psa_status_t status;
+	struct tfm_crypto_pack_iovec iov = {
+		.function_id = TFM_CRYPTO_CIPHER_ENCRYPT_SID,
+		.key_id = key,
+		.alg = alg,
+	};
+
+	psa_invec in_vec[] = {
+		{.base = &iov, .len = sizeof(struct tfm_crypto_pack_iovec)},
+		{.base = input, .len = input_length},
+	};
+	psa_outvec out_vec[] = {
+		{.base = output, .len = output_size}
+	};
+
+	status = API_DISPATCH(in_vec, out_vec);
+
+	*output_length = out_vec[0].len;
+	return status;
+}
+
+psa_status_t psa_cipher_decrypt(psa_key_id_t key,
+				psa_algorithm_t alg,
+				const uint8_t *input,
+				size_t input_length,
+				uint8_t *output,
+				size_t output_size,
+				size_t *output_length)
+{
+	psa_status_t status;
+	struct tfm_crypto_pack_iovec iov = {
+		.function_id = TFM_CRYPTO_CIPHER_DECRYPT_SID,
+		.key_id = key,
+		.alg = alg,
+	};
+
+	psa_invec in_vec[] = {
+		{.base = &iov, .len = sizeof(struct tfm_crypto_pack_iovec)},
+		{.base = input, .len = input_length},
+	};
+	psa_outvec out_vec[] = {
+		{.base = output, .len = output_size}
+	};
+
+	status = API_DISPATCH(in_vec, out_vec);
+
+	*output_length = out_vec[0].len;
+	return status;
+}
+
+psa_status_t psa_raw_key_agreement(psa_algorithm_t alg,
+				   psa_key_id_t private_key,
+				   const uint8_t *peer_key,
+				   size_t peer_key_length,
+				   uint8_t *output,
+				   size_t output_size,
+				   size_t *output_length)
+{
+	psa_status_t status;
+	struct tfm_crypto_pack_iovec iov = {
+		.function_id = TFM_CRYPTO_RAW_KEY_AGREEMENT_SID,
+		.alg = alg,
+		.key_id = private_key
+	};
+
+	psa_invec in_vec[] = {
+		{.base = &iov, .len = sizeof(struct tfm_crypto_pack_iovec)},
+		{.base = peer_key, .len = peer_key_length},
+	};
+
+	psa_outvec out_vec[] = {
+		{.base = output, .len = output_size},
+	};
+
+	status = API_DISPATCH(in_vec, out_vec);
+
+	*output_length = out_vec[0].len;
+
+	return status;
+}
+
+psa_status_t psa_key_derivation_setup(psa_key_derivation_operation_t *operation,
+				      psa_algorithm_t alg)
+{
+	struct tfm_crypto_pack_iovec iov = {
+		.function_id = TFM_CRYPTO_KEY_DERIVATION_SETUP_SID,
+		.alg = alg,
+		.op_handle = operation->handle,
+	};
+
+	psa_invec in_vec[] = {
+		{.base = &iov, .len = sizeof(struct tfm_crypto_pack_iovec)},
+	};
+	psa_outvec out_vec[] = {
+		{.base = &(operation->handle), .len = sizeof(uint32_t)},
+	};
+
+	return API_DISPATCH(in_vec, out_vec);
+}
+
+psa_status_t psa_key_derivation_set_capacity(psa_key_derivation_operation_t
+					     *operation,
+					     size_t capacity)
+{
+	struct tfm_crypto_pack_iovec iov = {
+		.function_id = TFM_CRYPTO_KEY_DERIVATION_SET_CAPACITY_SID,
+		.capacity = capacity,
+		.op_handle = operation->handle,
+	};
+
+	psa_invec in_vec[] = {
+		{.base = &iov, .len = sizeof(struct tfm_crypto_pack_iovec)},
+	};
+
+	return API_DISPATCH_NO_OUTVEC(in_vec);
+}
+
+psa_status_t psa_key_derivation_input_bytes(psa_key_derivation_operation_t
+					    *operation,
+					    psa_key_derivation_step_t step,
+					    const uint8_t *data,
+					    size_t data_length)
+{
+	struct tfm_crypto_pack_iovec iov = {
+		.function_id = TFM_CRYPTO_KEY_DERIVATION_INPUT_BYTES_SID,
+		.step = step,
+		.op_handle = operation->handle,
+	};
+
+	psa_invec in_vec[] = {
+		{.base = &iov, .len = sizeof(struct tfm_crypto_pack_iovec)},
+		{.base = data, .len = data_length},
+	};
+
+	return API_DISPATCH_NO_OUTVEC(in_vec);
+}
+
+psa_status_t psa_key_derivation_output_key(const psa_key_attributes_t
+					   *attributes,
+					   psa_key_derivation_operation_t
+					   *operation,
+					   psa_key_id_t *key)
+{
+	struct tfm_crypto_pack_iovec iov = {
+		.function_id = TFM_CRYPTO_KEY_DERIVATION_OUTPUT_KEY_SID,
+		.op_handle = operation->handle,
+	};
+
+	psa_invec in_vec[] = {
+		{.base = &iov, .len = sizeof(struct tfm_crypto_pack_iovec)},
+		{.base = attributes, .len = sizeof(psa_key_attributes_t)},
+	};
+
+	psa_outvec out_vec[] = {
+		{.base = key, .len = sizeof(psa_key_id_t)}
+	};
+
+	return API_DISPATCH(in_vec, out_vec);
+}
+
+psa_status_t psa_key_derivation_input_integer(psa_key_derivation_operation_t
+					      *operation,
+					      psa_key_derivation_step_t step,
+					      uint64_t value)
+{
+	struct tfm_crypto_pack_iovec iov = {
+		.function_id = TFM_CRYPTO_KEY_DERIVATION_INPUT_INTEGER_SID,
+		.step = step,
+		.value = value,
+		.op_handle = operation->handle,
+	};
+
+	psa_invec in_vec[] = {
+		{.base = &iov, .len = sizeof(struct tfm_crypto_pack_iovec)},
+	};
+
+	return API_DISPATCH_NO_OUTVEC(in_vec);
+}
+
+psa_status_t psa_key_derivation_verify_bytes(__unused
+					     psa_key_derivation_operation_t
+					     *operation,
+					     __unused const uint8_t
+					     *expected_output,
+					     __unused size_t output_length)
+{
+	/* To be implemented when the PSA backend supports it */
+	return PSA_ERROR_NOT_SUPPORTED;
+}
+
+psa_status_t psa_key_derivation_verify_key(__unused
+					   psa_key_derivation_operation_t
+					   *operation,
+					   __unused psa_key_id_t expected)
+{
+	/* To be implemented when the PSA backend supports it */
+	return PSA_ERROR_NOT_SUPPORTED;
+}
+
+/* The implementation of the following helper function is marked
+ * weak to allow for those integrations where this is directly
+ * provided by the psa_crypto_client.c module of Mbed TLS
+ */
+__weak
+void psa_reset_key_attributes(psa_key_attributes_t *attributes)
+{
+	memset(attributes, 0, sizeof(*attributes));
+}
diff --git a/core/arch/arm/plat-stm32mp2/psa_service/tfm_crypto_defs.h b/core/arch/arm/plat-stm32mp2/psa_service/tfm_crypto_defs.h
new file mode 100644
index 000000000..2c008d233
--- /dev/null
+++ b/core/arch/arm/plat-stm32mp2/psa_service/tfm_crypto_defs.h
@@ -0,0 +1,187 @@
+/* SPDX-License-Identifier: BSD-3-Clause */
+/*
+ * Copyright (c) 2018-2022, Arm Limited. All rights reserved.
+ * Copyright (c) 2025, STMicroelectronics
+ *
+ */
+
+#ifndef __TFM_CRYPTO_DEFS_H__
+#define __TFM_CRYPTO_DEFS_H__
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#include "psa/crypto.h"
+#ifdef PLATFORM_DEFAULT_CRYPTO_KEYS
+#include "crypto_keys/tfm_builtin_key_ids.h"
+#else
+#include "tfm_builtin_key_ids.h"
+#endif /* PLATFORM_DEFAULT_CRYPTO_KEYS */
+
+/**
+ * \brief The maximum supported length of a nonce through the TF-M
+ *        interfaces
+ */
+#define TFM_CRYPTO_MAX_NONCE_LENGTH (16u)
+
+/**
+ * \brief This type is used to overcome a limitation in the number of maximum
+ *        IOVECs that can be used especially in psa_aead_encrypt and
+ *        psa_aead_decrypt. By using this type we pack the nonce and the actual
+ *        nonce_length at part of the same structure
+ *
+ */
+struct tfm_crypto_aead_pack_input {
+	uint8_t nonce[TFM_CRYPTO_MAX_NONCE_LENGTH];
+	uint32_t nonce_length;
+};
+
+/**
+ * \brief Structure used to pack non-pointer types in a call to PSA Crypto APIs
+ *
+ */
+struct tfm_crypto_pack_iovec {
+	psa_key_id_t key_id;     /*!< Key id */
+	psa_algorithm_t alg;     /*!< Algorithm */
+	uint32_t op_handle;      /*!< Client context handle associated to a
+				  *   multipart operation
+				  */
+	uint32_t ad_length;        /*!< Additional Data length for multipart AEAD */
+	uint32_t plaintext_length; /*!< Plaintext length for multipart AEAD */
+
+	struct tfm_crypto_aead_pack_input aead_in; /*!< Packs AEAD-related inputs */
+
+	uint16_t function_id;    /*!< Used to identify the function in the
+				  *   API dispatcher to the service backend
+				  *   See tfm_crypto_func_SID, for detail
+				  */
+	uint16_t step;           /*!< Key derivation step */
+	union {
+		uint32_t capacity;     /*!< Key derivation capacity */
+		uint64_t value;      /*!< Key derivation integer for update*/
+	};
+};
+
+/**
+ * \brief Type associated to the group of a function encoding. There can be
+ *        nine groups (Random, Key management, Hash, MAC, Cipher, AEAD,
+ *        Asym sign, Asym encrypt, Key derivation).
+ */
+enum tfm_crypto_group_id_t {
+	TFM_CRYPTO_GROUP_ID_RANDOM          = UINT8_C(1),
+	TFM_CRYPTO_GROUP_ID_KEY_MANAGEMENT  = UINT8_C(2),
+	TFM_CRYPTO_GROUP_ID_HASH            = UINT8_C(3),
+	TFM_CRYPTO_GROUP_ID_MAC             = UINT8_C(4),
+	TFM_CRYPTO_GROUP_ID_CIPHER          = UINT8_C(5),
+	TFM_CRYPTO_GROUP_ID_AEAD            = UINT8_C(6),
+	TFM_CRYPTO_GROUP_ID_ASYM_SIGN       = UINT8_C(7),
+	TFM_CRYPTO_GROUP_ID_ASYM_ENCRYPT    = UINT8_C(8),
+	TFM_CRYPTO_GROUP_ID_KEY_DERIVATION  = UINT8_C(9)
+};
+
+#define BVAL(x) (((uint16_t)((((uint16_t)(x)) << 8) & 0xFF00)) - 1)
+
+/**
+ * \brief This type defines numerical progressive values
+ * identifying a function API exposed through the interfaces
+ * (S or NS). It's used to dispatch the requests
+ * from S/NS to the corresponding API implementation in
+ * the Crypto service backend.
+ *
+ * \note Each function SID is encoded as uint16_t.
+ *        +------------+------------+
+ *        |  Group ID  |  Func ID   |
+ *        +------------+------------+
+ *   (MSB)15         8 7          0(LSB)
+ *
+ */
+enum tfm_crypto_func_sid_t {
+	BASE__RANDOM         = BVAL(TFM_CRYPTO_GROUP_ID_RANDOM),
+	TFM_CRYPTO_GENERATE_RANDOM_SID,
+	BASE__KEY_MANAGEMENT = BVAL(TFM_CRYPTO_GROUP_ID_KEY_MANAGEMENT),
+	TFM_CRYPTO_GET_KEY_ATTRIBUTES_SID,
+	TFM_CRYPTO_OPEN_KEY_SID,
+	TFM_CRYPTO_CLOSE_KEY_SID,
+	TFM_CRYPTO_IMPORT_KEY_SID,
+	TFM_CRYPTO_DESTROY_KEY_SID,
+	TFM_CRYPTO_EXPORT_KEY_SID,
+	TFM_CRYPTO_EXPORT_PUBLIC_KEY_SID,
+	TFM_CRYPTO_PURGE_KEY_SID,
+	TFM_CRYPTO_COPY_KEY_SID,
+	TFM_CRYPTO_GENERATE_KEY_SID,
+	BASE__HASH           = BVAL(TFM_CRYPTO_GROUP_ID_HASH),
+	TFM_CRYPTO_HASH_COMPUTE_SID,
+	TFM_CRYPTO_HASH_COMPARE_SID,
+	TFM_CRYPTO_HASH_SETUP_SID,
+	TFM_CRYPTO_HASH_UPDATE_SID,
+	TFM_CRYPTO_HASH_CLONE_SID,
+	TFM_CRYPTO_HASH_FINISH_SID,
+	TFM_CRYPTO_HASH_VERIFY_SID,
+	TFM_CRYPTO_HASH_ABORT_SID,
+	BASE__MAC            = BVAL(TFM_CRYPTO_GROUP_ID_MAC),
+	TFM_CRYPTO_MAC_COMPUTE_SID,
+	TFM_CRYPTO_MAC_VERIFY_SID,
+	TFM_CRYPTO_MAC_SIGN_SETUP_SID,
+	TFM_CRYPTO_MAC_VERIFY_SETUP_SID,
+	TFM_CRYPTO_MAC_UPDATE_SID,
+	TFM_CRYPTO_MAC_SIGN_FINISH_SID,
+	TFM_CRYPTO_MAC_VERIFY_FINISH_SID,
+	TFM_CRYPTO_MAC_ABORT_SID,
+	BASE__CIPHER         = BVAL(TFM_CRYPTO_GROUP_ID_CIPHER),
+	TFM_CRYPTO_CIPHER_ENCRYPT_SID,
+	TFM_CRYPTO_CIPHER_DECRYPT_SID,
+	TFM_CRYPTO_CIPHER_ENCRYPT_SETUP_SID,
+	TFM_CRYPTO_CIPHER_DECRYPT_SETUP_SID,
+	TFM_CRYPTO_CIPHER_GENERATE_IV_SID,
+	TFM_CRYPTO_CIPHER_SET_IV_SID,
+	TFM_CRYPTO_CIPHER_UPDATE_SID,
+	TFM_CRYPTO_CIPHER_FINISH_SID,
+	TFM_CRYPTO_CIPHER_ABORT_SID,
+	BASE__AEAD           = BVAL(TFM_CRYPTO_GROUP_ID_AEAD),
+	TFM_CRYPTO_AEAD_ENCRYPT_SID,
+	TFM_CRYPTO_AEAD_DECRYPT_SID,
+	TFM_CRYPTO_AEAD_ENCRYPT_SETUP_SID,
+	TFM_CRYPTO_AEAD_DECRYPT_SETUP_SID,
+	TFM_CRYPTO_AEAD_GENERATE_NONCE_SID,
+	TFM_CRYPTO_AEAD_SET_NONCE_SID,
+	TFM_CRYPTO_AEAD_SET_LENGTHS_SID,
+	TFM_CRYPTO_AEAD_UPDATE_AD_SID,
+	TFM_CRYPTO_AEAD_UPDATE_SID,
+	TFM_CRYPTO_AEAD_FINISH_SID,
+	TFM_CRYPTO_AEAD_VERIFY_SID,
+	TFM_CRYPTO_AEAD_ABORT_SID,
+	BASE__ASYM_SIGN      = BVAL(TFM_CRYPTO_GROUP_ID_ASYM_SIGN),
+	TFM_CRYPTO_ASYMMETRIC_SIGN_MESSAGE_SID,
+	TFM_CRYPTO_ASYMMETRIC_VERIFY_MESSAGE_SID,
+	TFM_CRYPTO_ASYMMETRIC_SIGN_HASH_SID,
+	TFM_CRYPTO_ASYMMETRIC_VERIFY_HASH_SID,
+	BASE__ASYM_ENCRYPT   = BVAL(TFM_CRYPTO_GROUP_ID_ASYM_ENCRYPT),
+	TFM_CRYPTO_ASYMMETRIC_ENCRYPT_SID,
+	TFM_CRYPTO_ASYMMETRIC_DECRYPT_SID,
+	BASE__KEY_DERIVATION = BVAL(TFM_CRYPTO_GROUP_ID_KEY_DERIVATION),
+	TFM_CRYPTO_RAW_KEY_AGREEMENT_SID,
+	TFM_CRYPTO_KEY_DERIVATION_SETUP_SID,
+	TFM_CRYPTO_KEY_DERIVATION_GET_CAPACITY_SID,
+	TFM_CRYPTO_KEY_DERIVATION_SET_CAPACITY_SID,
+	TFM_CRYPTO_KEY_DERIVATION_INPUT_BYTES_SID,
+	TFM_CRYPTO_KEY_DERIVATION_INPUT_KEY_SID,
+	TFM_CRYPTO_KEY_DERIVATION_INPUT_INTEGER_SID,
+	TFM_CRYPTO_KEY_DERIVATION_KEY_AGREEMENT_SID,
+	TFM_CRYPTO_KEY_DERIVATION_OUTPUT_BYTES_SID,
+	TFM_CRYPTO_KEY_DERIVATION_OUTPUT_KEY_SID,
+	TFM_CRYPTO_KEY_DERIVATION_ABORT_SID,
+};
+
+/**
+ * \brief This macro is used to extract the group_id from an encoded function id
+ *        by accessing the upper 8 bits. A \a _function_id is uint16_t type
+ */
+#define TFM_CRYPTO_GET_GROUP_ID(_function_id) \
+	((enum tfm_crypto_group_id_t)(((uint16_t)(_function_id) >> 8) & 0xFF))
+
+#ifdef __cplusplus
+}
+#endif
+
+#endif /* __TFM_CRYPTO_DEFS_H__ */
diff --git a/core/arch/arm/plat-stm32mp2/sub.mk b/core/arch/arm/plat-stm32mp2/sub.mk
index c0b884d8c..c3269857a 100644
--- a/core/arch/arm/plat-stm32mp2/sub.mk
+++ b/core/arch/arm/plat-stm32mp2/sub.mk
@@ -3,6 +3,7 @@ global-incdirs-y += .
 srcs-y += main.c
 srcs-y += stm32mp_pm.c
 subdirs-y += drivers
+subdirs-$(CFG_STM32_PSA_SERVICE) += psa_service
 
 srcs-$(CFG_DISPLAY) += display.c
 srcs-$(CFG_SCMI_SCPFW) += scmi_server_scpfw.c
diff --git a/core/arch/arm/plat-stm32mp2/tfm_builtin_key_ids.h b/core/arch/arm/plat-stm32mp2/tfm_builtin_key_ids.h
new file mode 100644
index 000000000..70debc00c
--- /dev/null
+++ b/core/arch/arm/plat-stm32mp2/tfm_builtin_key_ids.h
@@ -0,0 +1,29 @@
+/* SPDX-License-Identifier: BSD-3-Clause */
+/*
+ * Copyright (c) 2022-2023, Arm Limited. All rights reserved.
+ *
+ */
+
+#ifndef __TFM_BUILTIN_KEY_IDS_H__
+#define __TFM_BUILTIN_KEY_IDS_H__
+
+/**
+ * \brief The persistent key identifiers for TF-M builtin keys.
+ *
+ * \note The value of TFM_BUILTIN_KEY_ID_MIN (and therefore of the whole range)
+ * is completely arbitrary except for being inside the PSA builtin keys range.
+ * The range is specified by the limits defined through
+ * MBEDTLS_PSA_KEY_ID_BUILTIN_MIN and MBEDTLS_PSA_KEY_ID_BUILTIN_MAX
+ */
+enum tfm_builtin_key_id_t {
+	TFM_BUILTIN_KEY_ID_MIN = 0x7FFF815Bu,
+	TFM_BUILTIN_KEY_ID_HUK,
+	TFM_BUILTIN_KEY_ID_IAK,
+#ifdef TFM_PARTITION_DELEGATED_ATTESTATION
+	TFM_BUILTIN_KEY_ID_DAK_SEED,
+#endif /* TFM_PARTITION_DELEGATED_ATTESTATION */
+	TFM_BUILTIN_KEY_ID_PLAT_SPECIFIC_MIN = 0x7FFF816Bu,
+	TFM_BUILTIN_KEY_ID_MAX = 0x7FFF817Bu,
+};
+
+#endif /* __TFM_BUILTIN_KEY_IDS_H__ */
-- 
2.47.2

